pkg/kv/kvserver/replica_write.go

```diff
@@ -345,52 +316,52 @@ func (r *Replica) executeWriteBatch(
 	}
 }
 
-func rangeUnavailableMessage(
-	s *redact.StringBuilder,
+func rangeUnavailableError(
 	desc *roachpb.RangeDescriptor,
+	replDesc roachpb.ReplicaDescriptor,
 	lm liveness.IsLiveMap,
 	rs *raft.Status,
-	ba *roachpb.BatchRequest,
-	dur time.Duration,
-) {
-	var liveReplicas, otherReplicas []roachpb.ReplicaDescriptor
+) error {
+	nonLiveRepls := roachpb.MakeReplicaSet(nil)
 	for _, rDesc := range desc.Replicas().Descriptors() {
 		if lm[rDesc.NodeID].IsLive {
-			liveReplicas = append(liveReplicas, rDesc)
-		} else {
-			otherReplicas = append(otherReplicas, rDesc)
+			continue
 		}
+		nonLiveRepls.AddReplica(rDesc)
 	}
 
-	// Ensure that these are going to redact nicely.
-	var _ redact.SafeFormatter = ba
+	canMakeProgress := desc.Replicas().CanMakeProgress(
+		func(replDesc roachpb.ReplicaDescriptor) bool {
+			return lm[replDesc.NodeID].IsLive
+		},
+	)
+
+	// Ensure good redaction.
+	var _ redact.SafeFormatter = nonLiveRepls
 	var _ redact.SafeFormatter = desc
-	var _ redact.SafeFormatter = roachpb.ReplicaSet{}
-
-	s.Printf(`have been waiting %.2fs for proposing command %s.
-This range is likely unavailable.
-Please submit this message to Cockroach Labs support along with the following information:
-
-Descriptor:  %s
-Live:        %s
-Non-live:    %s
-Raft Status: %+v
-
-and a copy of https://yourhost:8080/#/reports/range/%d
-
-If you are using CockroachDB Enterprise, reach out through your
-support contract. Otherwise, please open an issue at:
-
-  https://github.com/cockroachdb/cockroach/issues/new/choose
-`,
-		dur.Seconds(),
-		ba,
-		desc,
-		roachpb.MakeReplicaSet(liveReplicas),
-		roachpb.MakeReplicaSet(otherReplicas),
-		redact.Safe(rs), // raft status contains no PII
-		desc.RangeID,
+	var _ redact.SafeFormatter = replDesc
+
+	err := errors.Errorf("replica %s of %s is unavailable", desc, replDesc)
+	err = errors.Wrapf(
+		err,
+		"raft status: %+v", redact.Safe(rs), // raft status contains no PII
 	)
+	if len(nonLiveRepls.AsProto()) > 0 {
+		err = errors.Wrapf(err, "replicas on non-live nodes: %v (lost quorum: %t)", nonLiveRepls, !canMakeProgress)
+	}
+
+	return err
+}
+
+func (r *Replica) rangeUnavailableError() error {
```

Comment:
nit: consider moving this elsewhere (preferably into `replica_circuit_breaker.go`) -- it is no longer relevant for this file.